_posts/2020-12-02-solidity-underhanded-contest-winners.md

```diff
@@ -0,0 +1,100 @@
+---
+layout: post
+published: true
+title: 'Announcing the Winners of the Solidity Underhanded Contest 👨‍💻🏅'
+date: '2020-12-02'
+author: Franziska Heintel
+category: Announcements
+---
+
+After thorough assessment of all submissions, we are happy to share the winners of this year's [Solidity Underhanded Contest](https://underhanded.soliditylang.org/)!
+
+Before we dive into the winning submissions, we'd like to thank all participants for taking part. In total, we received 16 submissions which you can find in [this repo](https://github.com/ethereum/solidity-underhanded-contest/tree/master/submissions_2020). All 16 submissions are eligible for a "qualified submission" Solidity Underhanded POAP NFT. We will be in touch with participants shortly with details on the claiming process of those.
+
+Now, let's have a look at this year's winners, starting from the 5th place!
+
+## Solidity Underhanded Contest 2020 Winners
+
+### 5️⃣ th Place: [Marius van der Wijden](https://github.com/ethereum/solidity-underhanded-contest/tree/master/submissions_2020/submission14) 
+
+_commentary by [Alex Beregszaszi](https://twitter.com/alexberegszaszi)_
+
+This submission highlights the need for clearly distinguishing error conditions from return values. The `ecrecover` precompile has a single output, the address, which is set to zero in case of errors. Many high-level wrappers for this precompile introduce new error conditions, but still reuse the same mechanic of returning a “zero address” on error. This was the case with OpenZeppelin’s version [prior](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v2.5.0/contracts/cryptography/ECDSA.sol) to the 3.0 series.
+
+Using this validation shortcoming, the submission allows anyone to create unlimited "vote accepted" transactions.
+
+Solidity has always supported multiple return values which allows returning a status/error code and data separately, but more importantly the established good practice is to use `require` statements for validation, which most `ecrecover` use cases follow today.
+
+While the submission is written clearly and the issue is not obvious without dissection both the `closeVote` and the *trusted* `ECDSA.recover` functions, in our opinion there are two shortcomings raising suspicion:
+1. Append only data structures are not commonly used and there have been various published bugs about clearing them.
```

Comment:
Did you take this "append only" from my comment?

Reply:
@axic 👆

Reply:
The `users` array is append only, as per `enter()`/`exit()`. The cleaning is a not a problem here, so we can remove that. And the problem is not 100% due to the append-only nature, but it plays part of it.

Reply:
Feel free to reword my entire section.